app/controllers/reviews_controller.rb

```diff
@@ -9,7 +9,8 @@ def create
     if @review.save
       redirect_to product_path(@product)
     else
-      redirect_to product_path(@product)
+      # redirect_to product_path(@product)
+      redirect_to product_path(@product), alert: "Cannot insert review #{@product.errors.full_messages.join(', ')}."
```

Comment:
When the review instance is failed to save, the errors message will be inside `@review.errors`, if you try to get the review's errors from `@product.errors` you will get nothing, and it will not show any error message in alert. Please fix this. Also, you should at "." after the message "Cannot insert review " to separate the first message with error messages.